### ISSUE TITLE
[BUG] Adding another project to cart fails if there's one in the cart already with qty = 1

### ISSUE THREAD

**Issue by kolaente** [integer-net/magento2-shippingpreselection#9]
Adding a product (Let's call it "product a") to cart with qty = 1 works fine. But when I then want to add another product ("product B") to the cart with some other qty, adding that second product fails with an error message "The requested qty is not available".

This is the quantity settings of product A:
![image](https://user-images.githubusercontent.com/13721712/140745933-de03ea74-6703-4bd8-be70-f81011605e60.png)

Product B has some some quantity > 1. Adding it individually (without Product A already in cart) works fine.

I dug into it a bit and it looks like this happens because this module saves the cart. That tries to set the quantity again for product A which won't work because it does not have enough quantity available at this point. Not quite sure if this is a core issue or only related to this module.

**Comment by kolaente** [comment]
The method called [here](https://github.com/integer-net/magento2-shippingpreselection/blob/master/src/Plugin/PreselectShippingMethod.php#L95) saves the quote. That's where it then fails.

**Comment by kolaente** [comment]
I'm not quite sure what would be a good way to fix this. Maybe plugging a different method than `getQuote`?

**Comment by amenk** [comment]
Wondering why it still fails to add product B, even the exception is caught ...

**Comment by kolaente** [comment]
Maybe because the exception rewinds the transaction?

**Comment by amenk** [comment]
You mean the exception rewineded the transaction before getting caught -- possible.
You could also try to comment-out
                $quote->addErrorInfo('error', null, $e->getCode(), $e->getMessage());
There is a setHasError() maybe that rewinds something

**Comment by kolaente** [comment]
No, the actual exception gets re-thrown a couple of times. I think it originally happens here and gets rolled back: https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/Model/ResourceModel/Db/AbstractDb.php#L433

**Comment by kolaente** [comment]
Uncommenting the ` $quote->addErrorInfo(` doesn't work either

**Comment by kolaente** [comment]
I'm actually not quite sure why magento tries to _add_ all existing products to cart when saving the cart (at vendor/magento/module-quote/Model/Quote.php:1804)

**Comment by kolaente** [comment]
I think if we could manage to set `reset_count` to the item's buy request we should be able to trigger resetting the qty [here](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Quote/Model/Quote/Item/Processor.php#L94). That would need to happen [here](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Quote/Model/Quote/Item/CartItemPersister.php#L79)

**Comment by kolaente** [comment]
Okay it does work when I edit the source [at that place here](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Quote/Model/Quote/Item/CartItemPersister.php#L79) to contain the if.

**Comment by kolaente** [comment]
This is the commit that fixes it: https://github.com/magento/magento2/commit/5478c03cfe34030ad377594c7e6398836d304477

**Comment by kolaente** [comment]
It looks like adding https://github.com/magento/magento2/commit/5478c03cfe34030ad377594c7e6398836d304477 via composer patches fixes this issue.

@schmengler @lbuchholz since there's no magento version yet with the patch it might make sense to add a note to the readme about a patch being necessary? Or patch it in the module directly?

There's probably other places as well where this bug is triggered but it looks a lot like an issue with this module (even though it is not, but we'd never noticed that bug without using this).
